File: src/main/java/dev/epicsquid/whoosh/utils/TeleportUtils.java
package dev.epicsquid.whoosh.utils;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.phys.Vec3;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.common.util.ITeleporter;
import net.minecraftforge.event.entity.EntityTeleportEvent;

import javax.annotation.Nullable;
import java.util.function.Function;

public class TeleportUtils {

	// Based on the old school Ender IO Travel Staff teleport code
	public static boolean shortTeleport(Level level, Player player, InteractionHand hand) {
		Vec3 targetVec = player.position().add(0, player.getEyeHeight(), 0);
		Vec3 lookVec = player.getLookAngle();
		BlockPos target = null;
		for (double i = 7; i >= 2; i -= 0.5) {
			Vec3 v3d = targetVec.add(lookVec.multiply(i, i, i));
			target = new BlockPos(Math.round(v3d.x), Math.round(v3d.y), Math.round(v3d.z));
			if (canTeleportTo(level, target.below())) {
				break;
			} else {
				target = null;
			}
		}
		if (target != null) {
			if (!player.getLevel().isClientSide) {
				Vec3 teleportVec = checkTeleport(player, target);
				if (teleportVec == null) {
					return false;
				}
				player.teleportTo(teleportVec.x(), teleportVec.y(), teleportVec.z());
			}
			player.fallDistance = 0;
			player.swing(hand, true);
			player.playNotifySound(SoundEvents.ENDERMAN_TELEPORT, SoundSource.PLAYERS, 1F, 1F);
			return true;
		}
		return false;
	}

	public static boolean canTeleportTo(BlockGetter level, BlockPos target) {
		return !level.getBlockState(target.immutable().above(1)).canOcclude()
						&& !level.getBlockState(target.immutable().above(2)).canOcclude()
						&& target.getY() >= level.getMinBuildHeight();
	}

	@Nullable
	private static Vec3 checkTeleport(Player player, BlockPos target) {
		EntityTeleportEvent event = new EntityTeleportEvent(player, target.getX() + 0.5, target.getY(), target.getZ() + 0.5);
		if (MinecraftForge.EVENT_BUS.post(event)) {
			return null;
		}
		return new Vec3(event.getTargetX(), event.getTargetY(), event.getTargetZ());
	}
}

File: src/main/java/dev/epicsquid/whoosh/utils/TeleportUtils.kt
//package dev.epicsquid.whoosh.utils
//
//import net.minecraft.core.BlockPos
//import net.minecraft.sounds.SoundEvents
//import net.minecraft.sounds.SoundSource
//import net.minecraft.world.InteractionHand
//import net.minecraft.world.entity.player.Player
//import net.minecraft.world.level.BlockGetter
//import net.minecraft.world.level.Level
//import net.minecraft.world.phys.Vec3
//import net.minecraftforge.common.MinecraftForge
//import net.minecraftforge.event.entity.EntityTeleportEvent
//
//object TeleportUtils {
//    // Based on the old school Ender IO Travel Staff teleport code
//    fun shortTeleport(level: Level, player: Player, hand: InteractionHand?): Boolean {
//        val targetVec = player.position().add(0.0, player.eyeHeight.toDouble(), 0.0)
//        val lookVec = player.lookAngle
//        var target: BlockPos? = null
//        var i = 7.0
//        while (i >= 2) {
//            val v3d = targetVec.add(lookVec.multiply(i, i, i))
//            target = BlockPos(Math.round(v3d.x).toDouble(), Math.round(v3d.y).toDouble(), Math.round(v3d.z).toDouble())
//            target = if (canTeleportTo(level, target.below())) {
//                break
//            } else {
//                null
//            }
//            i -= 0.5
//        }
//        if (target != null) {
//            if (!player.getLevel().isClientSide) {
//                val teleportVec = checkTeleport(player, target) ?: return false
//                player.teleportTo(teleportVec.x(), teleportVec.y(), teleportVec.z())
//            }
//            player.fallDistance = 0f
//            player.swing(hand, true)
//            player.playNotifySound(SoundEvents.ENDERMAN_TELEPORT, SoundSource.PLAYERS, 1f, 1f)
//            return true
//        }
//        return false
//    }
//
//    fun canTeleportTo(level: BlockGetter, target: BlockPos): Boolean {
//        return (!level.getBlockState(target.immutable().above(1)).canOcclude()
//                && !level.getBlockState(target.immutable().above(2)).canOcclude() && target.y >= level.minBuildHeight)
//    }
//
//    private fun checkTeleport(player: Player, target: BlockPos): Vec3? {
//        val event = EntityTeleportEvent(player, target.x + 0.5, target.y.toDouble(), target.z + 0.5)
//        return if (MinecraftForge.EVENT_BUS.post(event)) {
//            null
//        } else Vec3(event.targetX, event.targetY, event.targetZ)
//    }
//}
